### PR TITLE
2d noise kernel

### PR DESCRIPTION
**Pull request by Winnie08L**

### CODE REVIEW

**Comment by Winnie08L** [comment]
in build.py build_isotropic_kernels_frmo_function(), get attributeError if use the build_kernel_from function (the one defined previously for 1d) cuz the basis is not the right type, need to fix that

File: surface_potential_analysis/kernel/build.py [hunk 1]
```diff
@@ -95,7 +92,38 @@ def build_isotropic_kernel_from_function(
...
+    ]
+        beta(x-x'), the correllation as a function of displacement
+
+    Returns
+    -------
+    IsotropicNoiseKernel[
+        TupleBasisWithLengthLike[*tuple[FundamentalPositionBasis[Any, Any], ...]],
+    ]
+    """
+    return tuple(build_isotropic_kernel_from_function(basis_i, fn) for basis_i in basis)
```
Comment (Matt-Ord) [inline surface_potential_analysis/kernel/build.py:126]: add a build_isotropic_kernel() which works on a flat axis to sort these type issues, also add a build_isotropic_kernel_stacked_from_function which gets the full stacked kernel like build_isotropic_kernel_from_function currently does

Comment (Matt-Ord) [inline surface_potential_analysis/kernel/build.py:126]: StackedBasis is not necessarily iterable!

File: surface_potential_analysis/kernel/gaussian.py [hunk 2]
```diff
@@ -88,6 +89,46 @@ def fn(
     return build_isotropic_kernel_from_function(basis, fn)
 
 
+def get_separate_gaussian_isotropic_noise_kernel(
```
Comment (Matt-Ord) [inline surface_potential_analysis/kernel/gaussian.py:92]: should this be part of the public interface? Probably not...

File: surface_potential_analysis/kernel/kernel.py [hunk 3]
```diff
@@ -470,3 +475,27 @@ def get_isotropic_kernel_from_operators(
         as_diagonal_noise_operators_from_full(operators),
         assert_isotropic=assert_isotropic,
     )
+
+
+def outer_product(*arrays):
```
Comment (Matt-Ord) [inline surface_potential_analysis/kernel/kernel.py:480]: I don't think this should be part of the public interface

File: surface_potential_analysis/kernel/kernel.py [hunk 4]
```diff
@@ -470,3 +475,27 @@ def get_isotropic_kernel_from_operators(
...
+    grids = np.meshgrid(*arrays, indexing="ij")
+    return np.prod(grids, axis=0)
+
+
+def get_full_kernel_from_independent(
```
Comment (Matt-Ord) [inline surface_potential_analysis/kernel/kernel.py:485]: move to build.py. Be more consistent in naming these kernels - are they 'inependent kernels' or 'seperate kernels'. rename get_separate_gaussian_isotropic_noise_kernel accordingly. I would maybe call them something like axis kernels

File: surface_potential_analysis/kernel/plot.py [hunk 5]
```diff
@@ -537,3 +544,27 @@ def plot_isotropic_noise_kernel_2d_x(  # noqa: PLR0913
         kernel["basis"], kernel["data"], axes, idx, ax=ax, scale=scale, measure=measure
     )
     return fig, ax, mesh
+
+
+def _get_noise_kernel_percentage_error(
```
Comment (Matt-Ord) [inline surface_potential_analysis/kernel/plot.py:549]: needs a rebase i think?

**Review by Matt-Ord — CHANGES_REQUESTED** [review]

**Review by Matt-Ord — APPROVED** [review]
Well done! Was all pretty much working.

**Pull request merged**